libs/runtime/ebpf_epoch.c

```diff
@@ -4,130 +4,171 @@
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_PROPOSE_EPOCH_COMPLETE, ///< This message is sent only to CPU 0 to signal that epoch
+                                                        ///< computation is complete.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_EXIT_EPOCH, ///< This message is used when a thread running with IRQL < DISPATCH calls
+                                            ///< ebpf_epoch_exit on a different CPU than ebpf_epoch_enter. It is sent
+                                            ///< during from the thread that is existing the epoch to the CPU that
+                                            ///< ebpf_epoch_enter was called on.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_RUNDOWN_IN_PROGRESS, ///< This message is sent to each CPU to notify it that epoch code
+                                                     ///< is shutting down and that no future timers should be armed and
+                                                     ///< future messages should be ignored.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_IS_FREE_LIST_EMPTY,  ///< This message is sent to each CPU to query if it's local free
+                                                     ///< list is empty.
+} ebpf_epoch_cpu_message_type_t;
+
+/**
+ * @brief Message sent between CPUs.
+ */
+typedef struct _ebpf_epoch_cpu_message
+{
+    ebpf_epoch_cpu_message_type_t message_type;
+    union
+    {
+        struct
+        {
+            uint64_t current_epoch;          ///< The new current epoch.
+            uint64_t proposed_release_epoch; ///< Minimum epoch of all threads on the CPU.
+        } propose_epoch;
+        struct
+        {
+            uint64_t released_epoch; ///< The newest epoch that can be released.
+        } commit_epoch;
+        struct
+        {
+            ebpf_epoch_state_t* epoch_state; ///< Epoch state to remove.
+        } exit_epoch;
+        struct
+        {
+            uint8_t unused; ///< Unused.
+        } rundown_in_progress;
+        struct
+        {
+            bool is_empty; ///< True if the free list is empty.
+        } is_free_list_empty;
+    } message;
+    KEVENT completion_event; ///< Event to signal when the operation is complete.
+} ebpf_epoch_cpu_message_t;
+
+/**
+ * @brief Timer used to schedule epoch computation.
  */
-static volatile int64_t _ebpf_current_epoch = 1;
+static KTIMER _ebpf_epoch_compute_release_epoch_timer;
 
 /**
- * @brief _ebpf_release_epoch indicates the newest inactive epoch. All memory
- * free operations performed prior to this value can be safely deleted.
+ * @brief Message used to compute the release epoch.
  */
-static volatile int64_t _ebpf_release_epoch = 0;
+static ebpf_epoch_cpu_message_t _ebpf_epoch_compute_release_epoch_message = {0};
 
 /**
- * @brief Timer used to update _ebpf_release_epoch.
+ * @brief DPC used to process timer expiration.
  */
-static ebpf_timer_work_item_t* _ebpf_flush_timer = NULL;
+static KDPC _ebpf_epoch_timer_dpc;
```

Comment:
KDPC is a kernel-specific term.  Ok for this PR, but we'll want to change this to be a cxplat type once we create non-preemptible work item APIs in cxplat.

Reply:
Sure. The code uses the following APIs:
KeInitializeDpc -> Doesn't fail
KeSetTargetProcessorDpc -> Target which processor the DPC will run on.
KeInsertQueueDpc -> Queue the DPC to run.